File: ghost-coach-be/src/main/java/com/playmotech/ghostcoach/chat/ChatRepository.java
package com.playmotech.ghostcoach.chat;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface ChatRepository extends JpaRepository<ChatMessage, Long> {

    Page<ChatMessage> findBySessionIdOrderByCreatedAtAsc(Long sessionId, Pageable pageable);

    // Sliding-window memory: fetch recent messages strictly BEFORE the just-saved user message.
    // Returns DESC for efficient LIMIT; service reverses to chronological order for the prompt.
    List<ChatMessage> findBySessionIdAndIdLessThanOrderByIdDesc(
            Long sessionId, Long beforeId, Pageable pageable);
}

File: ghost-coach-mobile/lib/data/repository/chat_repository.dart
import 'package:ghost_coach/data/model/chat/chat_message.dart';
import 'package:ghost_coach/data/source/chat_remote_source.dart';

class ChatRepository {
  ChatRepository(this._remote);

  final ChatRemoteSource _remote;

  Future<List<ChatMessage>> history(int sessionId) async {
    final page = await _remote.history(sessionId);
    return page.content;
  }

  Future<List<ChatMessage>> send(int sessionId, String message) =>
      _remote.send(sessionId, message);
}
